File: src/com/codingbaby/ohmyidea/OhTypedActionHandler.java
package com.codingbaby.ohmyidea;

import com.intellij.codeInsight.lookup.Lookup;
import com.intellij.codeInsight.lookup.LookupManager;
import com.intellij.openapi.actionSystem.DataContext;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.editor.actionSystem.TypedActionHandler;
import org.jetbrains.annotations.NotNull;

import javax.swing.*;

/**
 */
public class OhTypedActionHandler implements TypedActionHandler {

    private final TypedActionHandler origHandler;
    @NotNull
    private final KeyHandler handler;

    public OhTypedActionHandler(TypedActionHandler origHandler) {
        this.origHandler = origHandler;
        handler = KeyHandler.getInstance();
    }


    @Override
    public void execute(@NotNull final Editor editor, final char charTyped, @NotNull final DataContext dataContext) {

        if (isEnabled(editor)) {
            // Run key handler outside of the key typed command for creating our own undoable commands
            SwingUtilities.invokeLater(new Runnable() {
                @Override
                public void run() {
                    try {
                        handler.handleKey(editor, KeyStroke.getKeyStroke(charTyped), dataContext);
                    } catch (Throwable e) {
                        e.printStackTrace();
                    }
                }
            });
        } else {
            origHandler.execute(editor, charTyped, dataContext);
        }
    }

    private boolean isEnabled(@NotNull Editor editor) {
        if (OhPlugin.isEnabled()) {
            final Lookup lookup = LookupManager.getActiveLookup(editor);
            return lookup == null || !lookup.isFocused();
        }
        return false;
    }
}

File: src/main/java/com/codingbaby/ohmyidea/OhTypedActionHandler.kt
package com.codingbaby.ohmyidea


import `fun`.codecode.OhPlugin
import com.intellij.codeInsight.lookup.LookupManager
import com.intellij.openapi.actionSystem.DataContext
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.editor.actionSystem.TypedActionHandler
import javax.swing.KeyStroke
import javax.swing.SwingUtilities

/**
 * 当编辑器发生文本输入的时候，由这个action来响应
 * 此action内部维护了系统的handler，当自己不处理的时候就交给系统的handler
 */
class OhTypedActionHandler(private val origHandler: TypedActionHandler) : TypedActionHandler {

    override fun execute(editor: Editor, charTyped: Char, dataContext: DataContext) {

        if (!OhPlugin.instance.active) {
            origHandler.execute(editor, charTyped, dataContext)
            return
        }

        if (!isEnabled(editor)) {
            origHandler.execute(editor, charTyped, dataContext)
            return
        }


        if (OhPlugin.instance.status !== EditorStatus.Insert) {
            SwingUtilities.invokeLater {
                KeyHandler.handleKey(editor, KeyStroke.getKeyStroke(charTyped), dataContext)
            }
            return
        }

        origHandler.execute(editor, charTyped, dataContext)

    }

    private fun isEnabled(editor: Editor): Boolean {
        val lookup = LookupManager.getActiveLookup(editor)
        return lookup == null || !lookup.isFocused
    }
}
